File: scripts/molab-rename.py
#!/usr/bin/env python3
"""Rename MoLab workspace notebooks to human titles from NOTEBOOKS.yaml."""

from __future__ import annotations

import argparse
import json
import subprocess
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[1]
REGISTRY = REPO_ROOT / "notebooks" / "trading" / "NOTEBOOKS.yaml"
MANIFEST = REPO_ROOT / ".molab-import" / "rename-manifest.json"
SCRIPTS = REPO_ROOT / "scripts"


def _load_registry() -> list[dict]:
    projects: list[dict] = []
    current: dict | None = None
    for raw in REGISTRY.read_text(encoding="utf-8").splitlines():
        line = raw.rstrip()
        if line.startswith("  - slug:"):
            if current:
                projects.append(current)
            current = {"slug": line.split(":", 1)[1].strip()}
            continue
        if current and line.startswith("    ") and ":" in line:
            key, value = line.strip().split(":", 1)
            current[key] = value.strip()
    if current:
        projects.append(current)
    return projects


def cmd_rename(args: argparse.Namespace) -> None:
    selected = _load_registry()
    if args.slug:
        selected = [p for p in selected if p["slug"] == args.slug]
    notebooks = []
    for p in selected:
        slug = p["slug"]
        notebooks.append(
            {
                "slug": slug,
                "title": p.get("title", slug),
                "catalog_url": p.get("molab_url", ""),
                "match_names": [slug, "notebook.py", slug.replace("-", " ")],
            }
        )
        print(f"will rename: {slug} -> {p.get('title', slug)}")

    MANIFEST.parent.mkdir(parents=True, exist_ok=True)
    MANIFEST.write_text(json.dumps({"notebooks": notebooks}, indent=2) + "\n", encoding="utf-8")

    subprocess.run(["bash", str(SCRIPTS / "molab-chrome.sh")], check=False)
    proc = subprocess.run(["node", str(SCRIPTS / "molab-rename.mjs"), "--manifest", str(MANIFEST)], cwd=SCRIPTS)
    if proc.returncode != 0:
        raise SystemExit("Rename failed. Try manual rename in MoLab (see docs/GITHUB.md).")


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--slug")
    sub = parser.add_subparsers(dest="command", required=True)
    p = sub.add_parser("rename")
    p.set_defaults(func=cmd_rename)
    args = parser.parse_args()
    args.func(args)


if __name__ == "__main__":
    main()

File: scripts/molab-rename.mjs
#!/usr/bin/env node
/**
 * Rename MoLab workspace notebooks to registry titles (not "notebook.py").
 */
import { readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { closeMolabBrowser, connectMolabBrowser, getOrCreateMolabPage } from "./molab-browser.mjs";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const MOLAB_HOME = "https://molab.marimo.io/notebooks";

async function parseArgs() {
  const args = process.argv.slice(2);
  let manifest = path.join(__dirname, "..", ".molab-import", "rename-manifest.json");
  let loginOnly = false;
  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--manifest" && args[i + 1]) manifest = path.resolve(args[++i]);
    else if (args[i] === "--login-only") loginOnly = true;
  }
  return { manifest, loginOnly };
}

async function waitForMolabHome(page) {
  await page.goto(MOLAB_HOME, { waitUntil: "domcontentloaded" });
  for (let i = 0; i < 120; i++) {
    if (
      page.url().includes("molab.marimo.io/notebooks") &&
      (await page.getByRole("button", { name: /new notebook/i }).isVisible().catch(() => false))
    ) {
      return;
    }
    await page.waitForTimeout(1000);
  }
  throw new Error("MoLab home not ready");
}

async function renameOne(page, entry) {
  const { title, catalog_url, slug, match_names = [] } = entry;
  console.log(`Renaming -> "${title}" (${slug})`);

  await waitForMolabHome(page);
  await page.waitForTimeout(800);

  const candidates = [title, slug, "notebook.py", ...match_names];
  let card = null;
  for (const name of candidates) {
    const loc = page.getByRole("link", { name: new RegExp(`^${name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}$`, "i") });
    if (await loc.count()) {
      card = loc.first();
      break;
    }
  }

  if (!card && catalog_url) {
    const nbId = catalog_url.split("/").pop();
    card = page.locator(`a[href*="${nbId}"]`).first();
  }

  if (!card || !(await card.count())) {
    throw new Error(`Could not find notebook card for ${slug}`);
  }

  await card.hover();
  await page.waitForTimeout(300);

  const row = card.locator("xpath=ancestor::*[self::div or self::li][1]");
  const menuBtn = row
    .locator('[aria-haspopup="menu"]')
    .or(row.getByRole("button", { name: /more|options|menu/i }))
    .last();
  if (await menuBtn.count()) {
    await menuBtn.click();
  } else {
    await card.click({ button: "right" });
  }
  await page.waitForTimeout(400);

  const renameItem = page
    .getByRole("menuitem", { name: /^rename$/i })
    .or(page.getByRole("menuitem").filter({ hasText: /^rename/i }));
  await renameItem.first().click({ timeout: 10_000 });

  const field = page
    .getByRole("textbox")
    .or(page.locator('input[type="text"]'))
    .filter({ hasNot: page.locator('[type="url"]') })
    .last();
  await field.fill(title);

  const save = page
    .getByRole("button", { name: /^save$/i })
    .or(page.getByRole("button", { name: /^rename$/i }))
    .or(page.getByRole("button", { name: /^ok$/i }));
  await save.first().click({ timeout: 10_000 });
  await page.waitForTimeout(1000);

  console.log(`  done`);
}

async function main() {
  const { manifest, loginOnly } = await parseArgs();
  const session = await connectMolabBrowser({ preferCdp: true });
  const page = await getOrCreateMolabPage(session.context);

  try {
    await waitForMolabHome(page);
    if (loginOnly) {
      console.log("Login OK. Run: ./scripts/molab-rename.sh");
      return;
    }

    const payload = JSON.parse(await readFile(manifest, "utf8"));
    for (const entry of payload.notebooks) {
      await renameOne(page, entry);
    }
  } finally {
    await closeMolabBrowser(session);
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
